lib/imagec/imagec.go

```diff
@@ -953,46 +988,72 @@ func (p *Pusher) GetReaderForLayer(layerID string) (*ArchiveStream, io.ReadClose
 	if err != nil {
 		msg := fmt.Sprintf("Failed to create tmp file: %s", err.Error())
 		log.Info(msg)
-		return nil, nil, fmt.Errorf(msg)
+		return "", 0, fmt.Errorf(msg)
 	}
 
-	log.Infof("Writing layer to temp file %s", f.Name())
+	tarSize, err := io.Copy(f, ar)
+	if err != nil {
+		f.Close()
+		os.Remove(f.Name())
+		msg := fmt.Sprintf("Failed to read from acrhive stream: %s", err.Error())
+		log.Info(msg)
+		return "", 0, fmt.Errorf(msg)
+	}
+
+	if err = f.Sync(); err != nil {
+		f.Close()
+		os.Remove(f.Name())
+		msg := fmt.Sprintf("Failed to flush tar file: %s", err.Error())
+		log.Info(msg)
+		return "", 0, fmt.Errorf(msg)
+	}
+	f.Close()
+
+	log.Infof("Wrote tar archive to %s, len = %d", f.Name(), tarSize)
+	return f.Name(), tarSize, nil
+}
+
+func (p *Pusher) writeCompressedTarFile(layerID string, tarFilePath string, tarSize int64, progressOutput progress.Output) (string, string, int64, error) {
+	tarFile, err := os.Open(tarFilePath)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to open tar file %s: %s", tarFilePath, err.Error())
+		log.Info(msg)
+		return "", "", 0, fmt.Errorf(msg)
+	}
+
+	gzPath := filepath.Join(os.TempDir(), layerID+".tar.gz")
+	gzFile, err := os.OpenFile(gzPath, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		msg := fmt.Sprintf("Failed to create gzipped tar file: %s", err.Error())
+		log.Info(msg)
+		return "", "", 0, fmt.Errorf(msg)
+	}
+
+	log.Infof("Writing compressed layer to temp file %s", gzFile.Name())
+
+	reader := progress.NewProgressReader(ioutils.NewCancelReadCloser(context.Background(), tarFile), progressOutput, tarSize, layerID, "Pushing")
```

Comment:
`layerID`-->`ShortID(layerID)`: docker uses the short id to update the progress output bar.